README

                  ## ### ### #   #   ### ###  #  # # # #
                  #   #  ##  #   #   ### ##   #  # # ###
                 ##   #  ### ### ### # # # #  #  ### # #

                By Fabien Chereau & al. : stellarium (at) free.fr
                    http://www.stellarium.org

Stellarium comes with ABSOLUTELY NO WARRANTY.  See the COPYING file for details.

--------------------------------------------------------------------------------

1. INTRODUCTION                  6. THANKS
2. INSTALLATION INSTRUCTIONS     7. NOTE
3. REQUIREMENT                   8. FULL REFERENCES & CREDITS
4. QUICK START                   9. LICENSE
5. CONFIGURATION                 10. APPENDIX

--------------------------------------------------------------------------------
1. INTRODUCTION
--------------------------------------------------------------------------------
Stellarium is a free GPL software which renders realistic skies in real time 
with openGL. It is available for Linux/Unix, Windows and MacOSX. With Stellarium,
you really see what you can see with your eyes, binoculars or a small telescope.

--------------------------------------------------------------------------------
2. INSTALLATION INSTRUCTIONS
--------------------------------------------------------------------------------
Please see the INSTALL file.

--------------------------------------------------------------------------------
3. REQUIREMENT
--------------------------------------------------------------------------------
Windows or linux/POSIX or CYGWIN or MACOSX environment
A 3d openGL acceleration card and a good CPU.

--------------------------------------------------------------------------------
4. QUICK START
--------------------------------------------------------------------------------
Navigation :
Use the direction keys or drag with the mouse to move the point of view.
Use page up and page down keys or mouse wheel to zoom in and out.
Use left mouse button to select an object, right button to select no object and
middle mouse button or SPACE to center on the selected object.
Zooming on nebulas or planets is very interesting....
Use J, K and L to increment/decrement time speed.
Press the H key for more help.

More help can be found on www.stellarium.org, and in the user guide.

--------------------------------------------------------------------------------
5. CONFIGURATION
--------------------------------------------------------------------------------
From version 0.6.0 the most useful configuration options can be set from within
the program.
You can however change it by hand in the file located in config/config.ini on
Windows/MacOSX and in $HOME/.stellarium/ on a unix system.

--------------------------------------------------------------------------------
6. THANKS
--------------------------------------------------------------------------------

	++ Special thanks to ++
	
	Rob Spearman (Digitalis Education Solutions, Inc.) for his help and financial 
    support.  Have a look at http://digitaliseducation.com if you want to buy a 
	planetarium projector running stellarium!

	Johan Meuris for his stunning drawings of the constellations.
	  (Johan is now an official member of the project)
		
	Johannes Gajdosik for his work on planets computation and projection.
	  (Johannes is now an official member of the project)
	
	Matthew Gates for his huge work on the User Guide.
	  (Matthew is now an official member of the project)
	 
	Nigel Kerr for his MacOSX port.
	 
	Fumio Yamamoto for his MacOSX packages, patches, support and documentation.
	
	Tony Furr for his work on the Graphical User Interface.

	Valery Lainey for his L1 and MarsSat ephemeris and his prompt an kind help
	with any question concerning planetary/satellite ephemeris.

	Norbert Zacharias from USNO for kindly sending a cut of the NOMAD catalogue
	on 4 DVD, answering NOMAD related questions, and finally copying the
	entiere NOMAD catalogue onto a disk.
    
	++ Thanks to ++
	
	Axel Mellinger who gave me the authorisation to use his impressive
	    milky way panorama.
	Jesse Jones for his callback library.
	The celestia project team from which I borrowed many code parts (GPL license).
	The libnova team for their planet calculation algorithms (GPL license).
	All the SDL team for their remarkable work.
	All the sourceforge team for hosting my project.
	Cedric Delfosse for his precious help on debian port.
	The GEPI team from the Observatoire Astronomique de Paris for 
		the great Hipparcos catalog.
    Brad Schaefer for his sky rendering algorithm.
    Sylvain Ferey for his optimisation of the grids drawing.
    Ben Wyatt (ben (at) wyatt100.freeserve.co.uk) for the Glpng library.
    Jean-Francois Tremblay for his porting on MacOSX.
	Vincent Caron for his parser bugfix and Linux compatibility bugfixes.
	Nick Porcino for his planet function.
    Tangui Morlier for his help on Linux System.
	Bill Gray (projectpluto.com) and Mark Huss (mark (at) mhuss.com) for all
	    the astro libraries.
    Chris Laurel (claurel (at) shatters.net) who makes Celestia.
	Yuuki Ninomiya (gm (at) debian.or.jp) for the parsecfg file.
	Antje Buchholz for her translation help and logistic support :)
	
	And... All the people who sent me mails and bug reports!


--------------------------------------------------------------------------------
7. NOTE
--------------------------------------------------------------------------------
    This program is free, but if you have nothing to do with your
    money, just go there and do something usefull with it :
    https://www.greenpeace.com/forms/gpicontr.html


--------------------------------------------------------------------------------
8. FULL REFERENCES & CREDITS
--------------------------------------------------------------------------------
1. Technical Articles
	1.1 The tone reproductor class
		The class mainly performs a fast implementation of the algorithm from
		the	paper [1], with more accurate values from [2]. The blue shift
		formula is taken from [3] and combined with the Scotopic vision
		formula from [4].
  		[1] "Tone Reproduction for Realistic Images", Tumblin and Rushmeier,
			IEEE Computer Graphics & Application, November 1993
  		[2] "Tone Reproduction and Physically Based Spectral Rendering",
			Devlin, Chalmers, Wilkie and Purgathofer in EUROGRAPHICS 2002
	  	[3] "Night Rendering", H. Wann Jensen, S. Premoze, P. Shirley,
			W.B. Thompson, J.A. Ferwerda, M.M. Stark
  		[4] "A Visibility Matching Tone Reproduction Operator for High Dynamic
			Range Scenes", G.W. Larson, H. Rushmeier, C. Piatko
	1.2 The skylight class
		The class is a fast implementation of the algorithm from the article
		"A Practical Analytic Model for Daylight" by A. J. Preetham, Peter
		Shirley and Brian Smits.
	1.3 The skybright class
		The class is a fast reimplementation of the VISLIMIT.BAS basic source
		code from Brad Schaefer's article on pages 57-60,  May 1998 _Sky &
		Telescope_,	"To the Visual Limits". The basic sources are available on
		the Sky and Telescope web site.
	1.4 Object luminance algorithm from GPL delphi program by Jan Hollan, 1999 (C)
		N.Copernicus Observatory and Planetarium in Brno.
		See http://svetlo.astro.cz/compute/lum.php

2. Included source code
	2.1 Iniparser library Copyright (c) 2000 by Nicolas Devillard.
	2.2 The stellastro and stellplanet libraries are mainly subsets of the
		libnova library (GPL) by Liam Girdwood (liam (at) nova-ioe.org)
	2.3 The orbit.cpp/h and solve.h files are directly borrowed from
		Celestia (Chris Laurel). (GPL license)
	2.4 Other pieces of code and ideas are from Celestia too (GPL license).
    2.5 Callback library by Jess Jones, see section 10.4 for full license.
    2.6 Several implementations of IMCCE theories for planet and satellite
        movement by Johannes Gajdosik (MIT-style license,
        see the corresponding files for the license text)

3. Data
	3.1 The Hipparcos star catalog
		From ESA (European Space Agency) and the Hipparcos mission.
		ref. ESA, 1997, The Hipparcos and Tycho Catalogues, ESA SP-1200
		http://cdsweb.u-strasbg.fr/ftp/cats/I/239
	3.2 The solar system data mainly comes from IMCCE and partly from Celestia.
	3.3 Polynesian constellations are based on diagrams from the Polynesian 
		Voyaging Society
	3.4 Chinese constellations are based on diagrams from the Hong Kong Space Museum
	3.5 Egyptian constellations are based on the work of Juan Antonio Belmonte, 
		Instituto de Astrofisica de Canarias
	3.6 The Tycho-2 Catalogue of the 2.5 Million Brightest Stars
		Hog E., Fabricius C., Makarov V.V., Urban S., Corbin T.,
		Wycoff G., Bastian U., Schwekendiek P., Wicenec A.
		<Astron. Astrophys. 355, L27 (2000)>
        http://cdsweb.u-strasbg.fr/ftp/cats/I/259
	3.7 Naval Observatory Merged Astrometric Dataset (NOMAD) version 1
		http://www.nofs.navy.mil/nomad
		Norbert Zacharias writes:
		"There are no fees, both UCAC and NOMAD are freely available
		with the only requirement that the source of the data
		(U.S. Naval Observatory) and original product name need to be
		provided with any distribution, as well as a description about
		any changes made to the data, if at all."
		The changes made to the data are:
		-) try to compute visual magnitude and color from the b,v,r values
		-) compute nr_of_measurements = the number of valid b,v,r values
		-) throw away or keep stars (depending on magnitude,
		   nr_of_measurements, combination of flags, tycho2 number)
		-) add all stars from hiparcos (incl. component solutions),
		   and tycho2+1st supplement
		-) reorganize the stars in several brigthness levels and triangular
		   zones according to position and magnitude
		The programs that are used to generate the star files are called
		"MakeCombinedCatalogue", "ParseHip", "ParseNomad",
		and can be found in the util subdirectory in source code.
		The position, magnitudes, and proper motions of the stars
		coming from NOMAD are unchanged, except for a possible
		loss of precision, especially in magnitude. When there is no
		v-magnitude, it is estimated from r or b magnitude.
		When there is no b- or v- magnitude, the color B-V is estimated from
		the other magnitudes.
		Also proper motions of faint stars are neglected at all.

4. Graphics
	4.1 All graphics are copyrighted by Fabien Chereau (GPL) except the ones
		mentioned below :
	4.2 The "earthmap" texture was created by NASA (Reto Stockli, NASA Earth 
		Observatory) using data from the MODIS instrument aboard the 
		Terra satellite (Public Domain). See chapter 10.1 for full credits.
	4.4 Moon texture map is Courtesy USGS Astrogeology Research Program,
		http://astrogeology.usgs.gov. (Public Domain, DFSG-free)
  	4.5 All other planet maps from David Seal's site: http://maps.jpl.nasa.gov/
	    see license in section 10.3
	4.6 The snowy landscape textures are from the GPL game tuxracer.
	    "Copyright 1999, 2000, 2001 Jasmin F. Patry" available under GPL license.
	4.7 The fullsky milky way panorama is created by Axel Mellinger, University
		of Potsdam, Germany. Further information and more pictures available from
		http://home.arcor-online.de/axel.mellinger/
		License: permission given to "Modify and redistribute this image if proper 
		credit to the original image is given."
	4.8 All messiers nebula pictures except those mentioned below from the Grasslands 
		Observatory : "Images courtesy of Tim Hunter and James McGaha, Grasslands 
		Observatory at http://www.3towers.com."
		License: permission given to "use the image freely" (including right to 
		modify and redistribute) "as long as it is credited."
	4.9 M31, and the Pleiades pictures come from Herm Perez :
		http://home.att.net/~hermperez/default.htm
		License: "Feel free to use these images, if you use them in a commercial 
		setting please attribute the source."
	4.10 Images of M8, M20, M33, NGC253, NGC1499, NGC2244, IC343 from Jean-Pierre Bousquet
	4.11 Images of M1, M15, M16, M27, M57, M97, NGC6946 from Stephane Dumont
	4.12 Images of M17, M44, NGC856, NGC884 from Maxime Spano
	4.13 Constellation art, GUI buttons, logo created by Johan Meuris (Jomejome) 
   		(jomejome at users.sourceforge.net ) http://users.pandora.be/jomejom/
		License: released under the Free Art License 
		(http://artlibre.org/licence.php/lalgb.html)
	4.14 The "earth-clouds" texture includes imagery owned by NASA.  
	See NASA's Visible Earth project at http://visibleearth.nasa.gov/
	License: 1. The imagery is free of licensing fees
		 2. NASA requires that they be provided a credit as the owners of the 
		    imagery
        The cloud texturing was taken from Celestia (GPL), 
        http://www.shatters.net/celestia/
   	
--------------------------------------------------------------------------------
9. LICENSE
--------------------------------------------------------------------------------
   Copyright (C) 2004-2006 Fabien Chereau fab.chereau (at) free.fr

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License
   as published by the Free Software Foundation; either version 2
   of the License, or (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

   See the COPYING file for more information regarding the GNU General
   Public License.

--------------------------------------------------------------------------------
10 APPENDIX :
--------------------------------------------------------------------------------
1. Full credits for image 4.2 :
	Author: Reto Stockli, NASA Earth Observatory, rstockli (at) climate.gsfc.nasa.gov
	Address of correspondance:
		Reto Stockli               
		ETH/IAC (NFS Klima) & NASA/GSFC Code 913 (SSAI)
		University Irchel
		Building 25 Room J53  
		Winterthurerstrasse 190        
		8057 Zurich, Switzerland  
	Phone:  +41 (0)1 635 5209     	
	Fax:    +41 (0)1 362 5197
	Email:  rstockli (at) climate.gsfc.nasa.gov
	http://earthobservatory.nasa.gov
	http://www.iac.ethz.ch/staff/stockli
	Supervisors: 
		Fritz Hasler and David Herring, NASA/Goddard Space Flight Center
	Funding:
		This project was realized under the SSAI subcontract 2101-01-027 (NAS5-01070)

	License :
		"Any and all materials published on the Earth Observatory are freely available 
	for re-publication or re-use, except where copyright is indicated."
	
	
2. License for the GLpng library
	/*
	* PNG loader library for OpenGL v1.45 (10/07/00)
	* by Ben Wyatt ben@wyatt100.freeserve.co.uk
	* Using LibPNG 1.0.2 and ZLib 1.1.3
	*
	* This software is provided 'as-is', without any express or implied warranty.
	* In no event will the author be held liable for any damages arising from the
	* use of this software.
	*
	* Permission is hereby granted to use, copy, modify, and distribute this
	* source code, or portions hereof, for any purpose, without fee, subject to
	* the following restrictions:
	*
	* 1. The origin of this source code must not be misrepresented. You must not
	*    claim that you wrote the original software. If you use this software in
	*    a product, an acknowledgment in the product documentation would be
	*    appreciated but is not required.
	* 2. Altered versions must be plainly marked as such and must not be
	*    misrepresented as being the original source.
	* 3. This notice must not be removed or altered from any source distribution.
	*/

	
 3. License for the JPL planets images (http://www.jpl.nasa.gov/images/policy/index.cfm)
	---
	Unless otherwise noted, images and video on JPL public web sites (public sites 
	ending with a jpl.nasa.gov address) may be used for any purpose without prior 
	permission, subject to the special cases noted below. Publishers who wish to 
	have authorization may print this page and retain it for their records; JPL 
	does not issue image permissions on an image by image basis. 
	By electing to download the material from this web site the user agrees: 
	1. that Caltech makes no representations or warranties with respect to 
	ownership of copyrights in the images, and does not represent others who may 
	claim to be authors or owners of copyright of any of the images, and makes no 
	warranties as to the quality of the images. Caltech shall not be responsible 
	for any loss or expenses resulting from the use of the images, and you 
	release and hold Caltech harmless from all liability arising from such use. 
	2. to use a credit line in connection with images. Unless otherwise noted in 
	the caption information for an image, the credit line should be "Courtesy 
	NASA/JPL-Caltech." 
	3. that the endorsement of any product or service by Caltech, JPL or NASA must 
	not be claimed or implied. 
	Special Cases: 
	* Prior written approval must be obtained to use the NASA insignia logo (the 
	blue "meatball" insignia), the NASA logotype (the red "worm" logo) and the 
	NASA seal. These images may not be used by persons who are not NASA employees 
	or on products (including Web pages) that are not NASA sponsored. In 
	addition, no image may be used to explicitly or implicitly suggest 
	endorsement by NASA, JPL or Caltech of commercial goods or services. Requests 
	to use NASA logos may be directed to Bert Ulrich, Public Services Division, 
	NASA Headquarters, Code POS, Washington, DC 20546, telephone (202) 358-1713, 
	fax (202) 358-4331, email bert.ulrich@hq.nasa.gov. 
	* Prior written approval must be obtained to use the JPL logo (stylized JPL 
	letters in red or other colors). Requests to use the JPL logo may be directed 
	to the Television/Imaging Team Leader, Media Relations Office, Mail Stop 
	186-120, Jet Propulsion Laboratory, Pasadena CA 91109, telephone (818) 
	354-5011, fax (818) 354-4537. 
	* If an image includes an identifiable person, using the image for commercial 
	purposes may infringe that person's right of privacy or publicity, and 
	permission should be obtained from the person. NASA and JPL generally do not 
	permit likenesses of current employees to appear on commercial products. For 
	more information, consult the NASA and JPL points of contact listed above. 
	* JPL/Caltech contractors and vendors who wish to use JPL images in 
	advertising or public relation materials should direct requests to the 
	Television/Imaging Team Leader, Media Relations Office, Mail Stop 186-120, 
	Jet Propulsion Laboratory, Pasadena CA 91109, telephone (818) 354-5011, fax 
	(818) 354-4537. 
	* Some image and video materials on JPL public web sites are owned by 
	organizations other than JPL or NASA. These owners have agreed to make their 
	images and video available for journalistic, educational and personal uses, 
	but restrictions are placed on commercial uses. To obtain permission for 
	commercial use, contact the copyright owner listed in each image caption. 
	Ownership of images and video by parties other than JPL and NASA is noted in 
	the caption material with each image. 
	---
	
4. License for the callback library by Jesse Jones :
	(C) Copyright Jesse Jones 2000. Permission to copy, use, modify, sell
	and distribute this software is granted provided this copyright
	notice appears in all copies. This software is provided "as is" without
	express or implied warranty, and with no claim as to its suitability for
	any purpose.



NEWS
-----
 This file is not maintained anymore. Please read the news on the main website
-----

[07-nov-2004]
As you can see, Stellarium website has been totally rewritten, hope you like it this way. Send your comments in the feedback forum links.

[17-oct-2004]
Stellarium v0.6.1 released!
This new release includes the long awaited impressive constellation art images by Johan Meuris!
It also features the addition of a new immersive landscape skin, improved planets and satellites orbits and many other features. Thanks to all the developers, packagers and graphists who helped me to achieve this new version!

[07-oct-2004]
MacOSX package available for version 0.6.0. Big thanks to <a href="http://sourceforge.net/users/cmdtab/">Fumio Yamamoto</a> for providing this package.

[09-sep-2004]
Stellarium v0.6.0 for window update. I recompiled the program with a new version of MingW32 and it fixed the very slow rendering
of the atmosphere under windows. Windows version is therefore 5 time faster! If you have windows, please re-download version 0.6.0 from the download section.

[31-may-2004]
Stellarium v0.6.0 available! After more than one year of work, discover the new version with major improvements. 
A big thank to all contributors who helped me in this hard task, especially to Rob Spearman from <a href="http://digitaliseducation.com/index.html">
Digitalis Education</a> , who now sells special planetarium projectors running Stellarium! Please have a look at the updated features list for an overview 
of the new functions.

[16-january-2004]
Stellarium v0.5.2 for MacOSX available in the download section! A big thank to James Lowery who was the fastest to successfuly make this package.

[19-december-2003]
A macOSX packager is needed! If you know how to create a macOSX package from the source distribution of stellarium 0.5.2, please help me! The aim is to obtain a Stellarium-0.5.2.dmg file similiar to the one available for version 0.4.9 in the download section.

[27-august-2003]
The next version I am working on will have features designed for use in a planetarium dome : fisheye projection with a field of view up to 360 degree, use of fisheye textures mapped to make landscapes, automatic zoom on selected objects, a text interface etc... The reason of that is an american company called Digitalis Education which is paying me for adding such cool features (that of course everybody will benefit as it is opensource). You just have to wait a few more weeks and it will finally be released! I know that I say that in every post, but each time, the release date is closer... ;)

[24-june-2003]
In one month i will graduate from my university! I am therefore looking for a real job. If you need a good engineer for a scientific job, I am interested. I'm even more interested if you are located in Finland. Please have a look at my resume, and (hopefully!) contact me.

[12-june-2003]
It's now 2 months I'm working on a new version of stellarium. A lot of big improvements are on construction : I totally reorganized old part of the sources in a good object oriented way. I use now new algorithms for planets positions calculation and astronomical operations. The precision is really improved and sun eclipses, mercury transit are now visible. Jupiter satellites are here (beta) and i created a new faster and better atmosphere algorithm. This next version will hopefully be released in a few weeks!!

[21-may-2003]
I try to integrate more and more the project management with sourceforge. Please don't send reports anymore directly to my Email adress, see the section bug reports and feature requests for such things. The final goal is to enable new team members to join the project by making the informations available to everybody.

[21-may-2003]
Stellarium is curently in active development, and soon will be released a new "reorganized" version with global improvement of astronomical precision (eclipses etc...), simplification of the source code for a better future improvements. Please be patient!

[11-april-2003]
Stellarium v0.5.1 released : Use SDL instead of glut. Use a win32 installer. Fixed bugs and increase stability. Added a clickable map in the position configuration. Added the CTRL + Up/Down keys to control the zoom for macOSX users. I wait now for the linux packagers to make some rpms! 

[15-january-2003]
Stellarium v0.5.0 released : fixed some bugs, increased rendering speed, improved POSIX standard compliance for configuration files. 

[13-november-2002]
MacOSX binary package available!! Thank to Gunther Engelbrecht, you can now install stellarium easily on MacOSX. Find the link in the download section. 

[12-november-2002]
Mandrake RPM package available! Daniel Gruen sent me this package which runs fine on my computer. Download it in the download section. 

[18-october-2002]
I need people to help me creating the MacOSX binary version! If you own a mac and know how to do, you're welcome. 

[18-october-2002]
I'm currently looking for a 6 months job starting in january : if you work in a company which offers jobs for computer science student like me, please contact me!! You can find my resume there : cvfab.html, cvfab.pdf. 

[17-october-2002]
Update of this web site. The new page is more light and contains new screenshots. No more french version of this site is available : it takes me too much time to update it. 

[17-october-2002]
Stellarium v0.4.9 for windows released! I finally succeed to remove a bug which only occured on win32 plateform. 

[04-october-2002]
Stellarium v0.4.9 released! Now use Hipparcos stars catalogue, there are now 120 000 stars (10 000 before). MACOSX is now supported! Please update your if you have an old version! 


src/draw.h
/*
 * Stellarium
 * Copyright (C) 2002 Fabien Chereau
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

#ifndef __DRAW_H__
#define __DRAW_H__

#include <string>
#include <fstream>
#include "stellarium.h"
#include "s_font.h"
#include "projector.h"
#include "fader.h"
#include "translator.h"
#include "StelModule.h"

class STexture;
class Navigator;
class ToneReproductor;
class LoadingBar;

// Class which manages a grid to display in the sky
class SkyGrid
{
public:
	enum SKY_GRID_TYPE
	{
		EQUATORIAL,
		ALTAZIMUTAL
	};
	// Create and precompute positions of a SkyGrid
	SkyGrid(SKY_GRID_TYPE grid_type = EQUATORIAL, unsigned int _nb_meridian = 24, unsigned int _nb_parallel = 17,
	 double _radius = 1., unsigned int _nb_alt_segment = 18, unsigned int _nb_azi_segment = 50);
    virtual ~SkyGrid();
	void draw(const Projector* prj) const;
	void setFontSize(double newFontSize);
	void setColor(const Vec3f& c) {color = c;}
	const Vec3f& getColor() {return color;}
	void update(int delta_time) {fader.update(delta_time);}
	void set_fade_duration(float duration) {fader.set_duration((int)(duration*1000.f));}
	void setFlagshow(bool b){fader = b;}
	bool getFlagshow(void) const {return fader;}
	void set_top_transparancy(bool b) { transparent_top= b; }
private:
	unsigned int nb_meridian;
	unsigned int nb_parallel;
	double radius;
	unsigned int nb_alt_segment;
	unsigned int nb_azi_segment;
	bool transparent_top;
	Vec3f color;
	Vec3f** alt_points;
	Vec3f** azi_points;
	bool (Projector::*proj_func)(const Vec3d&, Vec3d&) const;
	double fontSize;
	SFont& font;
	SKY_GRID_TYPE gtype;
	LinearFader fader;
};


// Class which manages a line to display around the sky like the ecliptic line
class SkyLine
{
public:
	enum SKY_LINE_TYPE
	{
		EQUATOR,
		ECLIPTIC,
		LOCAL,
		MERIDIAN
	};
	// Create and precompute positions of a SkyGrid
	SkyLine(SKY_LINE_TYPE _line_type = EQUATOR, double _radius = 1., unsigned int _nb_segment = 48);
    virtual ~SkyLine();
	void draw(const Projector *prj,const Navigator *nav) const;
	void setColor(const Vec3f& c) {color = c;}
	const Vec3f& getColor() {return color;}
	void update(int delta_time) {fader.update(delta_time);}
	void set_fade_duration(float duration) {fader.set_duration((int)(duration*1000.f));}
	void setFlagshow(bool b){fader = b;}
	bool getFlagshow(void) const {return fader;}
	void setFontSize(double newSize);
private:
	double radius;
	unsigned int nb_segment;
	SKY_LINE_TYPE line_type;
	Vec3f color;
	Vec3f* points;
	bool (Projector::*proj_func)(const Vec3d&, Vec3d&) const;
	LinearFader fader;
	double fontSize;
	SFont& font;
};


// Class which manages the displaying of the Milky Way
class MilkyWay : public StelModule
{
public:
	MilkyWay();
    virtual ~MilkyWay();
  
 	///////////////////////////////////////////////////////////////////////////
	// Methods defined in the StelModule class
	virtual void init(const InitParser& conf, LoadingBar& lb);
	virtual string getModuleID() const {return "milkyway";}
	virtual double draw(Projector *prj, const Navigator *nav, ToneReproductor *eye);
	virtual void update(double deltaTime) {fader.update((int)(deltaTime*1000));}
	virtual void updateI18n() {;}
	virtual void updateSkyCulture(LoadingBar& lb) {;}
	
	//! Get Milky Way intensity
	float getIntensity() {return intensity;}
	//! Set Milky Way intensity
	void setIntensity(float aintensity) {intensity = aintensity;}
	void setTexture(const string& tex_file);
	void setColor(const Vec3f& c) { color=c;}
	void setFlagShow(bool b){fader = b;}
	bool getFlagShow(void) const {return fader;}
private:
	float radius;
	STexture* tex;
	Vec3f color;
	float intensity;
	float tex_avg_luminance;
	LinearFader fader;
};


class Draw
{
public:
	// Draw a point... (used for tests)
	static void drawPoint(float X,float Y,float Z);
};

#endif // __DRAW_H__
